unixfs/src/dir/walk.rs

```diff
@@ -0,0 +1,777 @@
+                let segment = FileSegment::first(bytes, !file_continues);
+
+                let state = if file_continues || next.is_some() {
+                    State::Unfinished(Self {
+                        current: self.current,
+                        next,
+                        pending: self.pending,
+                    })
+                } else {
+                    State::Last(self.current)
+                };
+
+                Ok(ContinuedWalk::File(segment, Item::from(state)))
+            },
+            UnixFsType::Metadata => todo!("metadata?"),
+            UnixFsType::Symlink => {
+                let contents = flat.data.Data.as_deref().unwrap_or_default();
+
+                let (cid, name, depth) = self.next.expect("continued without next");
+                self.current.as_symlink(cid, &name, depth, metadata);
+
+                let state = if let Some(next) = self.pending.pop() {
+                    State::Unfinished(Self {
+                        current: self.current,
+                        next: Some(next),
+                        pending: self.pending,
+                    })
+                } else {
+                    State::Last(self.current)
+                };
+
+                Ok(ContinuedWalk::Symlink(bytes, Item::from(state)))
+            },
+        }
+    }
+
+    fn skip_current_file(mut self) -> Skipped {
+        use InnerKind::*;
+        match &mut self.current.kind {
+            File(_, visit @ Some(_)) => {
+                visit.take();
+
+                if self.next.is_some() {
+                    Skipped(State::Unfinished(self))
+                } else {
+                    Skipped(State::Last(self.current))
+                }
+            },
+            Bucket(_) => todo!("we could skip shards as well by ... maybe?"),
+            ref x => todo!("how to skip {:?}", x),
+        }
+    }
+
+    // TODO: we could easily split an 'static value for a directory or bucket, which would pop all
+    // entries at a single level out to do some parallel walking, though the skipping could already
+    // be used to do that... Maybe we could return the filevisit on Skipped to save user from
+    // re-creating one? How to do the same for directories?
+}
+
+enum Either<A, B> {
+    Left(A),
+    Right(B),
+}
+
+impl<A, B> Iterator for Either<A, B>
+    where A: Iterator,
+          B: Iterator<Item = <A as Iterator>::Item>
+{
+    type Item = A::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        use Either::*;
+        match self {
+            Left(a) => a.next(),
+            Right(b) => b.next(),
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        use Either::*;
+        match self {
+            Left(a) => a.size_hint(),
+            Right(b) => b.size_hint(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct InnerEntry {
+    kind: InnerKind,
+    path: PathBuf,
```

Comment:
might be better to use `Vec<String>` or ... String as the windows pathbuf might be something else. need to remember to try this out. this came up during the path to bytes conversion.

Reply:
It seems that windows works as well. I am quite certain we should support path names with slashes in them but no test case.